### PR TITLE
Fixes #9 Bridge error when no file is selected

### LINKED ISSUE #9: Bridge error when no file is selected
The bridge throws an error when a file element is posted without selecting a file.

    $_FILES
    Array ( [file] => Array ( [name] => [type] => [tmp_name] => [error] => 4 [size] => 0 ) ) 
    error = 4 = UPLOAD_ERR_NO_FILE

    $symfonyRequest->files->all()
    Array ( [file] =>)

    Symfony\Bridge\PsrHttpMessage\Factory\DiactorosFactory::getFiles()
    Catchable Fatal Error: Argument 1 passed to    
    Symfony\Bridge\PsrHttpMessage\Factory\DiactorosFactory::getFiles() must be of the type array, 
    null given

The problem is that the Symfony request object contains a null for file elements if no file was selected.  One could argue that it should return an UploadedFile object with an error code of UPLOAD_ERR_NO_FILE (4) but obviously this is not something that would be considered for S2.7.  Don't know if it has been discussed for 3.x.

However, the PSR7 spec seems clear:

    interface ServerRequestInterface extends RequestInterface
    public function getUploadedFiles();
    * This method returns upload metadata in a normalized tree, with each leaf
    * an instance of Psr\Http\Message\UploadedFileInterface.

So for PSR7 we should return metadata even if a file was not actually uploaded.  Which is exactly what Zend\Diactoros\ServerRequestFactory::fromGlobals() does.

It's easy enough to update the bridge to match the Zend PSR7 behavior when no file has been selected.  I'll be glad to submit a pull request.  Just wanted to make sure that it would be accepted.

### PR DESCRIPTION
**Pull request by ahundiak**
| Q             | A
| ------------- | ---
| Bug fix?      | yes
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #9
| License       | MIT
| Doc PR        | n/a

### CODE REVIEW

**Comment by nicolas-grekas** [comment]
Please see tests failure (double space + parse error on 5.3)

**Comment by dunglas** [comment]
@stof suggested to fix Symfony directly. What is the selected solution?

**Comment by ahundiak** [comment]
I don't see how the S2.7 Request object could be modified without introducing a significant bc break.  This is an easy fix which won't impact anything even if the request 3.x object is fixed.  And I do want to use the bridge for my 2.3/2.7 projects.

**Comment by stof** [comment]
@ahundiak avoiding to put ``null`` in the array of files would not be a BC break IMO. It would be a bugfix, because the code currently does not respect its expected signature (and we would stop returning null in one of the error case, not add a null value, so it would not be a BC break)

**Comment by ahundiak** [comment]
@stof - From the cookbook: 

http://symfony.com/doc/current/cookbook/doctrine/file_uploads.html#basic-setup

    public function upload()
    {
      // the file property can be empty if the field is not required
      if (null === $this->getFile()) {
        return;
      }

I'm guessing that quite a few people are just checking for null and not bothering to check the error status.  One could of course argue that the file status should always be checked but I know I don't.

Understand that I'm not arguing against "fixing" the request object.  It's definitely the right thing to do in 3.x.  But as far as the bridge goes for existing code, I don't see the harm in handling nulls.

**Comment by stof** [comment]
@ahundiak this is about a file in the Doctrine entity. It has nothing to do with the HttpFoundation error checking (which is checked by the validator component during the form binding in the example you linked.

and if you work with HttpFoundation directly, you have to check the errors otherwise your code is totally broken.

**Comment by ahundiak** [comment]
@stof - I'll be glad to submit a pr against the file parameter bag object.  The current behavior is documented in a test:

    public function testShouldSetEmptyUploadedFilesToNull()
    {
        $bag = new FileBag(array('file' => array(
            'name' => '',
            'type' => '',
            'tmp_name' => '',
            'error' => UPLOAD_ERR_NO_FILE,
            'size' => 0,
        )));

        $this->assertNull($bag->get('file'));
    }

Easy enough to make the change.

**Comment by dunglas** [comment]
What do you think @symfony/deciders? Should we merge this PR or change the (tested) Behavior in HttpFoundation?

**Comment by dunglas** [comment]
@fabpot what do you think about this one?

**Comment by xabbuh** [comment]
Imho changing the behaviour in the `FileBag` class would be a BC break (people might rely on the fact that we return `null` elements to output error messages) and I don't see where the current code is violating the public interface. So I am :+1: for finishing this PR.

ping @symfony/deciders

**Comment by fabpot** [comment]
:+1: for this patch

File: Factory/DiactorosFactory.php [hunk 1]
```diff
@@ -86,6 +86,10 @@ private function getFiles(array $uploadedFiles)
         $files = array();
 
         foreach ($uploadedFiles as $key => $value) {
+            if ($value === null) {
```
Comment (Tobion) [inline Factory/DiactorosFactory.php]: YODA: null === $value

File: Tests/Factory/DiactorosFactoryTest.php [hunk 2]
```diff
@@ -161,4 +161,38 @@ public function testCreateResponseFromBinaryFile()
...
     }
+
+    public function testUploadErrNoFile()
+    {
+        $file = new UploadedFile(null, null, null, 0, UPLOAD_ERR_NO_FILE, true);
+        $this->assertEquals(0,$file->getSize());
+        $this->assertEquals(UPLOAD_ERR_NO_FILE,$file->getError());
+
+        // SplFile returns false on error
+        $this->assertEquals('boolean',gettype(($file->getSize())));
```
Comment (Tobion) [inline Tests/Factory/DiactorosFactoryTest.php]: not needed as the assertion below is enough

File: Tests/Factory/DiactorosFactoryTest.php [hunk 3]
```diff
@@ -161,4 +161,38 @@ public function testCreateResponseFromBinaryFile()
...
         $this->assertEquals('Binary', $psrResponse->getBody()->__toString());
     }
+
+    public function testUploadErrNoFile()
+    {
+        $file = new UploadedFile(null, null, null, 0, UPLOAD_ERR_NO_FILE, true);
+        $this->assertEquals(0,$file->getSize());
+        $this->assertEquals(UPLOAD_ERR_NO_FILE,$file->getError());
+
+        // SplFile returns false on error
```
Comment (Tobion) [inline Tests/Factory/DiactorosFactoryTest.php]: this should be the 3rd argument to assertFalse

File: Tests/Factory/DiactorosFactoryTest.php [hunk 4]
```diff
@@ -161,4 +161,38 @@ public function testCreateResponseFromBinaryFile()
 
         $this->assertEquals('Binary', $psrResponse->getBody()->__toString());
     }
+
+    public function testUploadErrNoFile()
+    {
+        $file = new UploadedFile(null, null, null, 0, UPLOAD_ERR_NO_FILE, true);
+        $this->assertEquals(0,$file->getSize());
+        $this->assertEquals(UPLOAD_ERR_NO_FILE,$file->getError());
```
Comment (Tobion) [inline Tests/Factory/DiactorosFactoryTest.php]: missing space after comma

File: Tests/Factory/DiactorosFactoryTest.php [hunk 5]
```diff
@@ -161,4 +161,38 @@ public function testCreateResponseFromBinaryFile()
...
+
+        // SplFile returns false on error
+        $this->assertEquals('boolean',gettype(($file->getSize())));
+        $this->assertFalse($file->getSize());
+
+        // This is an integer, oddly enough internally size is declared as a string
+        $this->assertTrue(is_int($file->getClientSize()));
```
Comment (Tobion) [inline Tests/Factory/DiactorosFactoryTest.php]: please use assertInternalType

File: Tests/Factory/DiactorosFactoryTest.php [hunk 6]
```diff
@@ -161,4 +161,38 @@ public function testCreateResponseFromBinaryFile()
...
+    {
+        $file = new UploadedFile(null, null, null, 0, UPLOAD_ERR_NO_FILE, true);
+        $this->assertEquals(0,$file->getSize());
+        $this->assertEquals(UPLOAD_ERR_NO_FILE,$file->getError());
+
+        // SplFile returns false on error
+        $this->assertEquals('boolean',gettype(($file->getSize())));
+        $this->assertFalse($file->getSize());
+
+        // This is an integer, oddly enough internally size is declared as a string
```
Comment (Tobion) [inline Tests/Factory/DiactorosFactoryTest.php]: this comment use useless here as you are not testing UploadedFile here, thus the assertion below is awkward as well

File: Tests/Factory/DiactorosFactoryTest.php [hunk 7]
```diff
@@ -161,4 +161,38 @@ public function testCreateResponseFromBinaryFile()
 
         $this->assertEquals('Binary', $psrResponse->getBody()->__toString());
     }
+
+    public function testUploadErrNoFile()
+    {
+        $file = new UploadedFile(null, null, null, 0, UPLOAD_ERR_NO_FILE, true);
```
Comment (Tobion) [inline Tests/Factory/DiactorosFactoryTest.php]: the $path parameter must be a string according to phpdoc, so passing null is invalid

**Comment by Tobion** [comment]
:+1: for finishing this patch.

File: Tests/Factory/DiactorosFactoryTest.php (hunk 6 above)
Comment (Tobion) [inline Tests/Factory/DiactorosFactoryTest.php]: symfony/symfony#18404

**Comment by weaverryan** [comment]
Ping @ahundiak - will you be able to make the suggested changes?

**Comment by ahundiak** [comment]
When I get time.  If someone else want to do then that would be fine with
me.  I know the importance of coding standards but it's hard to muster the
energy to follow all of the undocumented ones.

**Comment by Danielss89** [comment]
I've fixed the stuff mentioned in code-review, in this PR: https://github.com/Danielss89/psr-http-message-bridge/pull/1 to @ahundiak original branch. If he doesn't merge within a couple of days, i will create a new PR for this repo instead.

**Comment by weaverryan** [comment]
@Danielss89 it looks like your PR got closed? You can totally also make a new PR - if you include the original commits, that would be awesome and we can get this guy taken care of.

**Comment by Danielss89** [comment]
Opened a new PR also. See #23

**Comment by xabbuh** [comment]
closing in favour of #23

File: Factory/DiactorosFactory.php (hunk 1 above)
Comment (hexus) [inline Factory/DiactorosFactory.php]: I've always been curious why the Symfony source prefers `null` to come first. What's the reasoning?

Comment (dunglas) [inline Factory/DiactorosFactory.php]: https://en.wikipedia.org/wiki/Yoda_conditions

**Pull request closed without merging**